wicketd/src/artifacts/update_plan.rs

```diff
@@ -0,0 +1,1016 @@
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::io;
+use std::io::Read;
+use tufaceous_lib::HostPhaseImages;
+use tufaceous_lib::RotArchives;
+
+/// `UpdatePlanBuilder` mirrors all the fields of `UpdatePlan`, but they're all
+/// optional: it can be filled in as we read a TUF repository.
+/// [`UpdatePlanBuilder::build()`] will (fallibly) convert from the builder to
+/// the final plan.
+#[derive(Debug)]
+pub(super) struct UpdatePlanBuilder<'a> {
+    // fields that mirror `UpdatePlan`
+    system_version: SemverVersion,
+    gimlet_sp: BTreeMap<Board, ArtifactIdData>,
+    gimlet_rot_a: Option<ArtifactIdData>,
+    gimlet_rot_b: Option<ArtifactIdData>,
+    psc_sp: BTreeMap<Board, ArtifactIdData>,
+    psc_rot_a: Option<ArtifactIdData>,
+    psc_rot_b: Option<ArtifactIdData>,
+    sidecar_sp: BTreeMap<Board, ArtifactIdData>,
+    sidecar_rot_a: Option<ArtifactIdData>,
+    sidecar_rot_b: Option<ArtifactIdData>,
+    host_phase_1: Option<ArtifactIdData>,
+    trampoline_phase_1: Option<ArtifactIdData>,
+    trampoline_phase_2: Option<ArtifactIdData>,
+    host_phase_2_hash: Option<ArtifactHash>,
+    control_plane_hash: Option<ArtifactHash>,
+
+    // extra fields we use to build the plan
+    extracted_artifacts: ExtractedArtifacts,
+    log: &'a Logger,
+}
+
+impl<'a> UpdatePlanBuilder<'a> {
+    pub(super) fn new(
+        system_version: SemverVersion,
+        log: &'a Logger,
+    ) -> Result<Self, RepositoryError> {
+        let extracted_artifacts = ExtractedArtifacts::new(log)?;
+        Ok(Self {
+            system_version,
+            gimlet_sp: BTreeMap::new(),
+            gimlet_rot_a: None,
+            gimlet_rot_b: None,
+            psc_sp: BTreeMap::new(),
+            psc_rot_a: None,
+            psc_rot_b: None,
+            sidecar_sp: BTreeMap::new(),
+            sidecar_rot_a: None,
+            sidecar_rot_b: None,
+            host_phase_1: None,
+            trampoline_phase_1: None,
+            trampoline_phase_2: None,
+            host_phase_2_hash: None,
+            control_plane_hash: None,
+
+            extracted_artifacts,
+            log,
+        })
+    }
+
+    pub(super) fn add_artifact(
+        &mut self,
+        artifact_id: ArtifactId,
+        artifact_hash: ArtifactHash,
+        reader: io::BufReader<impl io::Read>,
+        by_id: &mut BTreeMap<ArtifactId, Vec<ArtifactHashId>>,
+        by_hash: &mut HashMap<ArtifactHashId, ExtractedArtifactDataHandle>,
+    ) -> Result<(), RepositoryError> {
+        // If we don't know this artifact kind, we'll still serve it up by hash,
+        // but we don't do any further processing on it.
+        let Some(artifact_kind) = artifact_id.kind.to_known() else {
+            return self.add_unknown_artifact(
+                artifact_id,
+                artifact_hash,
+                reader,
+                by_id,
+                by_hash,
+            );
+        };
+
+        match artifact_kind {
```

Comment:
It took me a while to understand the structure of this. I think one thing I got tripped up on is the "`(not trampoline!)`" and "`(not host!)`" comments. Could you add a comment explaining the general structure of this function?

Reply:
I added a bit of a comment and also removed the assertions (which were confusing at best as you pointed out in a later comment)